### DISCUSSION TITLE
SyncNode has no add_variable_type()

### DISCUSSION THREAD

**Discussion by roin93** [FreeOpcUa/opcua-asyncio#500] [Q&A] [answered]
Hello,
I was told to use the opcua-asyncio library instead of the old library.
The async functions give me a lot of headache, therefore I tried to use the sync wrapper from this library.
Many functions I know form the api of the old library are not working with this library despite the fact that somewhere it is mentions that most (!) of the api is the same.
For example I cannot add variable_types to a server.

> Traceback (most recent call last):
>   File "main.py", line 276, in <module>
>     main()
>   File "main.py", line 231, in main
>     datatype_node = server.get_node("i=93").add_variable_type("ns=3;i=6001", source_node.get_browse_name(), source_node.get_data_type())
> AttributeError: 'SyncNode' object has no attribute 'add_variable_type'

Is there a wiki to check which functions are supported between the two libraries and which functions are only available in the async version of this library?

**Comment by oroulet** [comment, marked as answer]
this is open-source so there not many people wanting to spend time on doc. yes there are some methods missing here and there but the sync api is very very simple: just ad the name and args fo the missing function there: https://github.com/FreeOpcUa/opcua-asyncio/blob/84b4a60f39c9d71e4c22b9c036611783d03de13d/asyncua/sync.py#L454 and it will work

**Reply by roin93** [reply]
That could help. Thank you! I will try that out and give a feedback.

**Reply by oroulet** [reply]
and send a merge request after adding the missing calls

**Reply by roin93** [reply]
I added add_variable_type() and it worked as expected. Thank you for that.
Also I added the add_reference() function. I know the part I try to convert to this library worked with the old library.
Now I get the following Error messages:

> Browse_name: QualifiedName(NamespaceIndex=3, Name='Par')
> WARNING:asyncua.server.address_space:AddNodesItem: Requested Browsename Par already exists in Parent Node. ParentID:NodeId(Identifier='V_"Mes"."SndData"."Par"', NamespaceIndex=3, NodeIdType=<NodeIdType.String: 3>) --- ItemId:NodeId(Identifier='V_"Mes"."SndData"."Par"[1]', NamespaceIndex=3, NodeIdType=<NodeIdType.String: 3>)
> Error:... "The browse name is not unique among nodes that share the same relationship with the parent."(BadBrowseNameDuplicated)
> WARNING:asyncua.client.ua_client.UaClient:close_session was called but connection is closed
> WARNING:asyncua.client.ua_client.UaClient:close_secure_channel was called but connection is closed
> WARNING:asyncua.client.ua_client.UaClient:disconnect_socket was called but connection is closed
> Traceback (most recent call last):
>   File "main_async_sync.py", line 106, in repeat_node
>     new_node = target_node.add_variable(node_id, browse_name, values, datatype=data_type)
>   File "/home/msa_user-m/.conda/envs/opcua_test/lib/python3.7/site-packages/asyncua-0.9.14-py3.7.egg/asyncua/sync.py", line 94, in wrapper
>     result = self.tloop.post(aio_func(*args, **kwargs))
>   File "/home/msa_user-m/.conda/envs/opcua_test/lib/python3.7/site-packages/asyncua-0.9.14-py3.7.egg/asyncua/sync.py", line 52, in post
>     return futur.result()
>   File "/home/msa_user-m/.conda/envs/opcua_test/lib/python3.7/concurrent/futures/_base.py", line 435, in result
>     return self.__get_result()
>   File "/home/msa_user-m/.conda/envs/opcua_test/lib/python3.7/concurrent/futures/_base.py", line 384, in __get_result
>     raise self._exception
>   File "/home/msa_user-m/.conda/envs/opcua_test/lib/python3.7/site-packages/asyncua-0.9.14-py3.7.egg/asyncua/common/node.py", line 676, in add_variable
>     return await create_variable(self, nodeid, bname, val, varianttype, datatype)
>   File "/home/msa_user-m/.conda/envs/opcua_test/lib/python3.7/site-packages/asyncua-0.9.14-py3.7.egg/asyncua/common/manage_nodes.py", line 106, in create_variable
>     await _create_variable(parent.server, parent.nodeid, nodeid, qname, var, datatype=datatype, isproperty=False)
>   File "/home/msa_user-m/.conda/envs/opcua_test/lib/python3.7/site-packages/asyncua-0.9.14-py3.7.egg/asyncua/common/manage_nodes.py", line 276, in _create_variable
>     results[0].StatusCode.check()
>   File "/home/msa_user-m/.conda/envs/opcua_test/lib/python3.7/site-packages/asyncua-0.9.14-py3.7.egg/asyncua/ua/uatypes.py", line 335, in check
>     raise UaStatusCodeError(self.value)
> asyncua.ua.uaerrors._auto.BadBrowseNameDuplicated: "The browse name is not unique among nodes that share the same relationship with the parent."(BadBrowseNameDuplicated)
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "main_async_sync.py", line 283, in <module>
>     main()
>   File "main_async_sync.py", line 245, in main
>     repeat_node(client, server, source_node, datatype_node)
>   File "main_async_sync.py", line 139, in repeat_node
>     repeat_node(client, server, child, new_node)
>   File "main_async_sync.py", line 139, in repeat_node
>     repeat_node(client, server, child, new_node)
>   File "main_async_sync.py", line 111, in repeat_node
>     new_node = target_node.add_reference(node_id, "i=47")
>   File "/home/msa_user-m/.conda/envs/opcua_test/lib/python3.7/site-packages/asyncua-0.9.14-py3.7.egg/asyncua/sync.py", line 94, in wrapper
>     result = self.tloop.post(aio_func(*args, **kwargs))
>   File "/home/msa_user-m/.conda/envs/opcua_test/lib/python3.7/site-packages/asyncua-0.9.14-py3.7.egg/asyncua/sync.py", line 52, in post
>     return futur.result()
>   File "/home/msa_user-m/.conda/envs/opcua_test/lib/python3.7/concurrent/futures/_base.py", line 435, in result
>     return self.__get_result()
>   File "/home/msa_user-m/.conda/envs/opcua_test/lib/python3.7/concurrent/futures/_base.py", line 384, in __get_result
>     raise self._exception
>   File "/home/msa_user-m/.conda/envs/opcua_test/lib/python3.7/site-packages/asyncua-0.9.14-py3.7.egg/asyncua/common/node.py", line 652, in add_reference
>     _check_results(results, len(params))
>   File "/home/msa_user-m/.conda/envs/opcua_test/lib/python3.7/site-packages/asyncua-0.9.14-py3.7.egg/asyncua/common/node.py", line 24, in _check_results
>     r.check()
>   File "/home/msa_user-m/.conda/envs/opcua_test/lib/python3.7/site-packages/asyncua-0.9.14-py3.7.egg/asyncua/ua/uatypes.py", line 335, in check
>     raise UaStatusCodeError(self.value)
> asyncua.ua.uaerrors._auto.BadTargetNodeIdInvalid: "The target node id does not reference a valid node."(BadTargetNodeIdInvalid)
> ERROR:asyncio:Task was destroyed but it is pending!

The parameter I tried to add is a Variable which has 512 components (in fact it is some kind of array).
I know in the old library it is working, in the new one it is not. Any ideas about that?

**Reply by oroulet** [reply]
In the new library we added a check to make sure two nodes with the same parents do not have the same BrowseName. in python-opcua it was possible while not allowed by spec. So it looks like your new node as the same name as an existing node. Chaneg the name

**Reply by roin93** [reply]
I copy nodes from a Siemens OPCUA Server.
There is a Node "Par" which has 512 Components (Variables). But the Parent-Node ("Par") has also a Reference called "HasModellingRule":ExposesItsArray

Can you tell me how I can check if it is that kind or node? Currently I try to differentiate by evaluating the following expression

```
        elif node_class == NodeClass.Variable:
            values = child.get_value()
            # data_type = child.get_data_type()
            # check if it is variable or property
            if client.get_node(child.read_type_definition()).read_browse_name().to_string() == '0:PropertyType':
                try:
                    print("IS PROPERTY")
                    new_node = target_node.add_property(node_id, browse_name, values, datatype=data_type)
                except ua.UaStatusCodeError as e:
                    print("Error_Property:...", e)
                    # new_node = target_node.add_reference(node_id, "i=46")
            elif client.get_node(child.read_type_definition()).read_browse_name().to_string() == "0:BaseDataVariableType":
                try:
                    new_node = target_node.add_variable(node_id, browse_name, values, datatype=data_type)
                except ua.UaStatusCodeError as e:
                    print("Error:...", e)

                    #### HasComponent to this target_node. Then we should be fine with the tree.
                    new_node = target_node.add_reference(node_id, "i=47")
                    print("Added a reference")
            else:
                # some complex variable type...
                print("Normal variable...")
                print(client.get_node(child.read_type_definition()).read_browse_name().to_string())
                try:
                    new_node = target_node.add_variable(node_id, browse_name, values, datatype=data_type)
                except ua.UaStatusCodeError as e:
                    print("Error:...", e)
```

**Reply by AndreasHeine** [reply]
https://github.com/FreeOpcUa/opcua-asyncio/blob/84b4a60f39c9d71e4c22b9c036611783d03de13d/asyncua/common/node.py#L378

```python

import asyncio
from asyncua import Client, ua

async def main():
    client = Client("opc.tcp://127.0.0.1:4840")
    await client.connect()
    refs = await client.get_node("ns=0;i=17635").get_references(refs=ua.ObjectIds.HasModellingRule)
    print(refs[0].DisplayName.Text) # for ref in refs -> check if "exposeitsarray"
    await client.disconnect()

asyncio.run(main())
```

> C:\Users\andre\Desktop\Neuer Ordner>C:/Users/andre/AppData/Local/Programs/Python/Python37/python.exe "c:/Users/andre/Desktop/Neuer Ordner/client.py"
> Optional
> 
> C:\Users\andre\Desktop\Neuer Ordner>

**Reply by roin93** [reply]
@oroulet 

>  the sync api is very very simple: just ad the name and args fo the missing function there:

Can you also tell me how I can add something like server.address_space to the SyncServer?

**Reply by oroulet** [reply]
I do not understant. what is server.address_space?

**Reply by roin93** [reply]
[server.address_space](https://python-opcua.readthedocs.io/en/latest/opcua.server.html#opcua.server.address_space.AddressSpace)
I thought I could use that for some manipulation of the Server Node structure.

**Comment by roin93** [comment]
Added a pull request #515
